### PR TITLE
Issue #2974: Fix EmptyLineSeparator check when class members are separated by comments

### PR DESCRIPTION
**Pull request by jjlharrison**
Hi, 

I've attempted to fix issue #2974. I think I've managed to capture all of the edge cases by parsing comments from the beginning of the file. The code is quite complex, but I think that it needs to be due to all of the possible edge cases.

My test case covers all by 2 conditional branches in the new code. This means that the coverage check is failing.

I'm not expecting this pull-request to be merged at this stage, but as someone who hasn't contributed to Checkstyle before, I think I need some help with how to cover these 2 conditions. They only occur with invalid Java source files that have a single trailing slash at the end of a line. I can't work out how to add an invalid Java source file without the compiler complaining in the IDE. I tried giving the source file the extension ".txt", but the `TreeWalker` filters out files not ending with ".java".

As far as I am aware, all of the other checks are passing.

Thanks

### CODE REVIEW

**Comment by jjlharrison** [comment]
FYI, the uncovered conditions are `lineHasNextChar()` on [EmptyLineSeparatorCheck:810](https://github.com/checkstyle/checkstyle/pull/5486/files#diff-f369e38e24af88245a4ff7ffea1cd55eR810) and [EmptyLineSeparatorCheck:826](https://github.com/checkstyle/checkstyle/pull/5486/files#diff-f369e38e24af88245a4ff7ffea1cd55eR826).

**Comment by rnveach** [comment]
@jjlharrison Sorry for delay.

Please read over https://github.com/checkstyle/checkstyle/wiki/PR-rules. Squash all commits to 1 and keep them at 1.

Please fix TC error while I look into code coverage and shippable.
> EmptyLineSeparatorCheck.java (1)
890: getFlags() 'return' of Collection field flags

**Comment by rnveach** [comment]
> uncovered conditions are lineHasNextChar() on EmptyLineSeparatorCheck:810

`lineHasNextChar()` never returns `false`. If we are always looking at a comment in this method, then the comment's start is always 2 characters `/*`.

> EmptyLineSeparatorCheck:826

`lineHasNextChar()` and `nextChar() == '/'` never returns `false`.
Comments can only be single line or multi-line. This method is called if the multi-line fails, so what else can this be but a single line comment?

To me it looks like you need to remove this extra code.

**Comment by jjlharrison** [comment]
@rnveach:
> Please read over https://github.com/checkstyle/checkstyle/wiki/PR-rules. Squash all commits to 1 and keep them at 1.

I was aware of this rule, it was my intention to squash the commits and rebase once the problems have all been resolved.

> This method is called if the multi-line fails, so what else can this be but a single line comment?
> To me it looks like you need to remove this extra code.

That is correct, those conditions can only occur with invalid Java source code. I'm happy to remove the conditions, but I was trying to gracefully handle invalid files as I wasn't sure what the rules were for handling such input.

**Comment by jjlharrison** [comment]
@rnveach: Ok, I think I've fixed the TC build, I've removed the redundant conditions, and I've squashed the commits. I'm hoping all of the checks will pass now.

**Comment by rnveach** [comment]
@jjlharrison You plan to finish this? travis is showing an exception with regression. Please check it out and add a test to confirm it is fixed.
````
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: 52
    at java.lang.String.charAt (String.java:658)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.nextChar (EmptyLineSeparatorCheck.java:874)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkStartComment (EmptyLineSeparatorCheck.java:811)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkStartStringLiteral (EmptyLineSeparatorCheck.java:802)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkEndComment (EmptyLineSeparatorCheck.java:789)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkEndStringLiteral (EmptyLineSeparatorCheck.java:773)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkQuoteEscape (EmptyLineSeparatorCheck.java:759)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.checkEmpty (EmptyLineSeparatorCheck.java:745)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.processChar (EmptyLineSeparatorCheck.java:735)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.processLine (EmptyLineSeparatorCheck.java:711)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck$EmptyLineSeparatorParser.processLines (EmptyLineSeparatorCheck.java:697)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.categorizeLines (EmptyLineSeparatorCheck.java:622)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.hasEmptyLineBefore (EmptyLineSeparatorCheck.java:558)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.processPackage (EmptyLineSeparatorCheck.java:422)
    at com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck.visitToken (EmptyLineSeparatorCheck.java:312)
````

**Comment by jjlharrison** [comment]
@rnveach: I'm not sure what has happened to the Travis build, the commit it [built](https://travis-ci.org/checkstyle/checkstyle/jobs/333949905) is 354f7a0, my latest commit is 428feec. Is there any way to re-trigger the Travis build so it picks up the latest commit?

**Comment by jjlharrison** [comment]
@rnveach: Awesome, found a way to test those conditions I was having problems testing with a valid file. I've change the test to reproduce the problem and fixed the bug 👌. Thanks.

**Comment by rnveach** [comment]
@jjlharrison pitest is failing on whitespace package. Please generate report and fix the missing mutation it is asking for.
Let me know if you have any issues.

**Comment by romani** [comment]
@jjlharrison , please finish PR, please rebase on latest master and we changed pitest validation a bit.

**Comment by rnveach** [comment]
@jjlharrison whitespace package in pitest is still failing.
> `@@ -1 +1 @@`
> `-EmptyLineSeparatorCheck.java.html:<td class='covered'><pre><span class='survived'> if (currentLineFlags.contains(LineFlag.STRING_LITERAL_OPEN)) {</span></pre></td></tr>`
> `+`
> `Actual:`
> `EmptyLineSeparatorCheck.java.html:<td class='covered'><pre><span  class='survived'>                if (currentLineFlags.contains(LineFlag.STRING_LITERAL_OPEN)) {</span></pre></td></tr>`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 1]
```diff
@@ -593,4 +610,276 @@ private static boolean isTypeField(DetailAST variableDef) {
...
+                checkStringLiteral();
+            }
+        }
+
+        /**
+         * Checks whether the current character starts or terminates a string literal.
+         */
+        private void checkStringLiteral() {
+            if (currentChar() == '"') {
+                if (currentLineFlags.contains(LineFlag.STRING_LITERAL_OPEN)) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: this is the line pitest is complaining about

**Comment by jjlharrison** [comment]
@romani @rnveach : Hope that does it. 🤞

**Comment by rnveach** [comment]
@jjlharrison you can run the commands locally to verify it does fix the issue.
TC is now failing.
> 654: STRING_LITERAL_OPEN Field has no usages.

**Comment by jjlharrison** [comment]
@rnveach: I am running the following locally to check:
```
man clean verify
mvn -e clean test org.pitest:pitest-maven:mutationCoverage -Ppitest-checks-whitespace
```

**Comment by rnveach** [comment]
@romani CI is passing now.

**Comment by romani** [comment]
@rnveach , please finish review of this PR, I will do review after you.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 2]
```diff
@@ -476,23 +478,25 @@ private boolean isViolatingEmptyLineBetweenFieldsPolicy(DetailAST detailAST) {
...
+            final List<EnumSet<LineFlag>> lineFlags = categorizeLines();
+            for (int i = token.getLineNo() - 2; i > 0 && !result; i--) {
+                final EnumSet<LineFlag> flags = lineFlags.get(i);
+                if (flags.contains(LineFlag.EMPTY)) {
+                    // 1 line above line being checked is empty.
+                    final EnumSet<LineFlag> previousLineFlags = lineFlags.get(i - 1);
+                    if (previousLineFlags.contains(LineFlag.EMPTY)) {
+                        // 2 lines above the line being checked
+                        // and after the previous token are empty.
+                        result = true;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: we should break out of the for loop as soon as we hit a result we are looking for.
there is nothing in this loop that will turn it back to `false`.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 3]
```diff
@@ -551,9 +555,22 @@ private boolean hasEmptyLineBefore(DetailAST token) {
...
-            result = CommonUtils.isBlank(lineBefore);
+            final List<EnumSet<LineFlag>> flags = categorizeLines();
+            for (int i = lineNo - 2; i >= 0; i--) {
+                final EnumSet<LineFlag> lineFlags = flags.get(i);
+                if (!lineFlags.contains(LineFlag.COMMENT)) {
+                    if (lineFlags.contains(LineFlag.EMPTY)) {
+                        // Line is empty.
+                        result = true;
+                    }
+                    if (lineFlags.contains(LineFlag.CODE)) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: shouldn't this be an `else if`?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 4]
```diff
@@ -593,4 +610,252 @@ private static boolean isTypeField(DetailAST variableDef) {
...
+        COMMENT_OPEN,
+
+        /**
+         * Line is empty. It doesn't contain any comments or code, but may contain whitespace.
+         */
+        EMPTY
+    }
+
+    /**
+     * Parses a Java source file to assign each line a set of flags based on whether they have code,
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: Why are we parsing the Java source? It seems like you are attempting a crude attempt at understanding the source code when we have the ANTLR to do this for us.
Why can't we use the AST to understand the source?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 2 above)
Comment (jjlharrison) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: @rnveach : I think I was doing this in a previous revision, but when I do the build fails with the Checkstyle violation:
```
SingleBreakOrContinue: Loops should not contain more than a single "break" or "continue" statement
```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java (hunk 4 above)
Comment (jjlharrison) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: I investigated that route, but it seemed to me that the AST didn't have understanding of whitespace, JavaDoc, comments, and empty lines. If it is possible to change the way the AST is built to do this, the I agree, that would be better. Would you be able to point me in the right direction?

Also, prior to my changes the empty lines were detected by checking the lines with `CommonUtils.isBlank`, not by analysing the AST. I assumed this was because it couldn't be done with the AST.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: > empty lines were detected by checking the lines with CommonUtils.isBlank, not by analysing the AST

Any empty line is an empty line. The util method is fine for this case. Emptyness could also be determined by looking for holes in the AST where nothing is on a specific line.

> AST didn't have understanding of whitespace, JavaDoc, comments, and empty lines

This check is comment aware, so it knows about comments and javadocs. Comments aren't part of the `getAcceptableTokens` list, so `visitToken` won't stop on them by default but they are still in the tree.
Empty lines can be determined using `CommonUtils.isBlank`. The AST contains the line number.
Whitespace would have to be determined by looking at the line's column position and checking what comes before/after it.

> Would you be able to point me in the right direction?

I can but you would need to ask a more specific question.
Be aware, that comments are placed in weird locations sometimes, so it might be best that you add comments to the acceptable list to stop on them.

Example from issue:
````
    // Method 2 must fail
    public static final void method2() { }

    |--METHOD_DEF -> METHOD_DEF [8:4]
    |  |--MODIFIERS -> MODIFIERS [8:4]
    |  |  |--SINGLE_LINE_COMMENT -> // [7:4]
    |  |  |  `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
    |  |  |--LITERAL_PUBLIC -> public [8:4]
    |  |  |--LITERAL_STATIC -> static [8:11]
````

You can use http://checkstyle.sourceforge.net/cmdline.html#Command_line_usage option `-T` to print the AST with the comments to under the tree more.

Comment (jjlharrison) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: @rnveach: Ok, I can look into this. I wish I had known about that before I had sunk hours into making all the CIs pass.

**Comment by romani** [comment]
@rnveach , please finish your review.

@jjlharrison , please confirm that you addressed all review items and just wait for review.

**Comment by rnveach** [comment]
@jjlharrison Please also make a comment when you have finished my items. We don't get notifications on new pushes to a PR.

**Comment by rnveach** [comment]
@romani Review items are still not finished.

**Comment by romani** [comment]
@jjlharrison , do you still want to finish this PR ?please finish.

**Comment by romani** [comment]
@jjlharrison , please confirm that you want to stop this PR. It is completely ok, but it will let somebody else to continue your work.
If no reply till end of month, I will mark this PR as abandoned.

**Comment by jjlharrison** [comment]
@romani: I intend to continue this when I have time, but I am quite busy at the moment, so if someone else wants to pick it up then they can go for it. Although, this issue has been open for over 2 years and had been inactive for a year when I started work on it, so it doesn't appear to high on anyones priority list.

**Comment by romani** [comment]
@jjlharrison , thanks a lot for reply, at least I know that you are still willing to continue. I just try to keep open PRs under control, let me mark this PR as abandoned, if you will not find time till end of June, I will close PR, but code still be referenced in issue , you or anybody else can continue even you are always welcome. I just need to keep PR under control.

**Comment by romani** [comment]
@jjlharrison , ping, please try to finish or at least share your problems.

**Comment by jjlharrison** [comment]
@romani: The problems are already documented in this pull-request.

As I hinted at before, this pull request would have been finished by now if my time hadn’t been wasted by being asked to complete mutation coverage only to then be told fix this a different way.

**Comment by romani** [comment]
>if my time hadn’t been wasted by being asked to complete mutation coverage

This is unfortunate, but there are case when all done and CIs are green,  but results of regression testing is not acceptable.

**Comment by jjlharrison** [comment]
@romani: I accept that, and I agree with stringent quality checks, but it’s very demotivating and wasteful of time when this could have been discovered earlier rather than just blindly asking for CIs to be fixed without checking that what was being fixed was right.

Also, this isn’t a regression testing issue, the fix works, but there is just an implementation that is more consistent with how checks are normally written.

**Comment by romani** [comment]
@jjlharrison , sorry that maintainers can not predict all problems at the beginning. We try to not read PRs that are failing because CI will lead contributor to better test code and in most cases it lead to conner cases that change solution completely. We can not read all versions of your implementation. But we do help when contributor ask questions and point to us that solution is not obvious and ....

Let's forget this unfortunate case. I will re-read your issue and code, we agree on solution first. Deal ?

**Comment by romani** [comment]
@jjlharrison , please help me understand why you selected way with "parsing" of java code.

What I see is ... this Check is comments aware module, https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java#L255 , so it mean that it traverse Tree with comments. So all that you need is already in ASTree.

example from issue:
```
$ cat Test.java 
public class Test {

    public static final void method1() { }



    // Method 2 must fail
    public static final void method2() { }



    /**
     * Method 3 must fail
     */
    public static final void method3() { }
}

$ java -jar checkstyle-8.10.1-all.jar -T Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--METHOD_DEF -> METHOD_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |   |--LITERAL_PUBLIC -> public [3:4]
    |   |   |--LITERAL_STATIC -> static [3:11]
    |   |   `--FINAL -> final [3:18]
    |   |--TYPE -> TYPE [3:24]
    |   |   `--LITERAL_VOID -> void [3:24]
    |   |--IDENT -> method1 [3:29]
    |   |--LPAREN -> ( [3:36]
    |   |--PARAMETERS -> PARAMETERS [3:37]
    |   |--RPAREN -> ) [3:37]
    |   `--SLIST -> { [3:39]
    |       `--RCURLY -> } [3:41]
    |--METHOD_DEF -> METHOD_DEF [8:4]
    |   |--MODIFIERS -> MODIFIERS [8:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [7:4]
    |   |   |   `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
    |   |   |--LITERAL_PUBLIC -> public [8:4]
    |   |   |--LITERAL_STATIC -> static [8:11]
    |   |   `--FINAL -> final [8:18]
    |   |--TYPE -> TYPE [8:24]
    |   |   `--LITERAL_VOID -> void [8:24]
    |   |--IDENT -> method2 [8:29]
    |   |--LPAREN -> ( [8:36]
    |   |--PARAMETERS -> PARAMETERS [8:37]
    |   |--RPAREN -> ) [8:37]
    |   `--SLIST -> { [8:39]
    |       `--RCURLY -> } [8:41]
    |--METHOD_DEF -> METHOD_DEF [15:4]
    |   |--MODIFIERS -> MODIFIERS [15:4]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [12:4]
    |   |   |   |--COMMENT_CONTENT -> *\n     * Method 3 must fail\n      [12:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [14:5]
    |   |   |--LITERAL_PUBLIC -> public [15:4]
    |   |   |--LITERAL_STATIC -> static [15:11]
    |   |   `--FINAL -> final [15:18]
    |   |--TYPE -> TYPE [15:24]
    |   |   `--LITERAL_VOID -> void [15:24]
    |   |--IDENT -> method3 [15:29]
    |   |--LPAREN -> ( [15:36]
    |   |--PARAMETERS -> PARAMETERS [15:37]
    |   |--RPAREN -> ) [15:37]
    |   `--SLIST -> { [15:39]
    |       `--RCURLY -> } [15:41]
    `--RCURLY -> } [16:0]
```

attention to (there is clear gap between line  3 and 7-8, comments over method can be located a bit dipper in tree, but this is stable behavior):
```
    |       `--RCURLY -> } [3:41]
    |--METHOD_DEF -> METHOD_DEF [8:4]
    |   |--MODIFIERS -> MODIFIERS [8:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [7:4]
    |   |   |   `--COMMENT_CONTENT ->  Method 2 must fail\n [7:6]
```

Why we can not simply go over tree and find a gap that is more than 1 ? in example above gap is 3.

Ok, make example a bit complicated
```
$ cat Test.java 
public class Test {

    public void method1() { }

    // single line comment

    /** Some comment with empty line

    */
    // comment over method2
    public void method2() { }
}

$ java -jar checkstyle-8.10.1-all.jar -T Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--METHOD_DEF -> METHOD_DEF [3:4]
    |   |--MODIFIERS -> MODIFIERS [3:4]
    |   |   `--LITERAL_PUBLIC -> public [3:4]
    |   |--TYPE -> TYPE [3:11]
    |   |   `--LITERAL_VOID -> void [3:11]
    |   |--IDENT -> method1 [3:16]
    |   |--LPAREN -> ( [3:23]
    |   |--PARAMETERS -> PARAMETERS [3:24]
    |   |--RPAREN -> ) [3:24]
    |   `--SLIST -> { [3:26]
    |       `--RCURLY -> } [3:28]
    |--METHOD_DEF -> METHOD_DEF [11:4]
    |   |--MODIFIERS -> MODIFIERS [11:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [5:4]
    |   |   |   `--COMMENT_CONTENT ->  single line comment\n [5:6]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [7:4]
    |   |   |   |--COMMENT_CONTENT -> * Some comment with empty line\n\n     [7:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [9:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [10:4]
    |   |   |   `--COMMENT_CONTENT ->  comment over method2\n [10:6]
    |   |   `--LITERAL_PUBLIC -> public [11:4]
    |   |--TYPE -> TYPE [11:11]
    |   |   `--LITERAL_VOID -> void [11:11]
    |   |--IDENT -> method2 [11:16]
    |   |--LPAREN -> ( [11:23]
    |   |--PARAMETERS -> PARAMETERS [11:24]
    |   |--RPAREN -> ) [11:24]
    |   `--SLIST -> { [11:26]
    |       `--RCURLY -> } [11:28]
    `--RCURLY -> } [12:0]
```

tracing a case:
```
    |--METHOD_DEF -> METHOD_DEF [3:4]
    ......
    |       `--RCURLY -> } [3:28]
    |--METHOD_DEF -> METHOD_DEF [11:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [5:4]
    |   |   |   `--COMMENT_CONTENT ->  single line comment\n [5:6]
    |   |   |--BLOCK_COMMENT_BEGIN -> /* [7:4]
    |   |   |   |--COMMENT_CONTENT -> * Some comment with empty line\n\n     [7:6]
    |   |   |   `--BLOCK_COMMENT_END -> */ [9:4]
    |   |   |--SINGLE_LINE_COMMENT -> // [10:4]
    |   |   |   `--COMMENT_CONTENT ->  comment over method2\n [10:6]
```
previous method in on line 3
next method is on line 11
11 - 3 = 8 lines are some where 
look at comments
there is smth on line 5 ---- so line 4 is EMPTY !
theres is smth on line 7  ...... but nothing is on line 6 , so 6 is EMPTY !
block from line 7 finishes at line 9 , so line 8 is empty but is inside some block - skip
there is smth on line 10, ok
method is on line 11, ok
SO there is 2 EMPTY lines

looks simple , please let me know what I miss.

**Comment by jjlharrison** [comment]
> So all that you need is already in ASTree.

@romani: Yes, I know the correct way to fix it, @rnveach [told me this back in March](https://github.com/checkstyle/checkstyle/pull/5486#discussion_r175460079).

> please help me understand why you selected way with "parsing" of java code

This fix I have submitted was mearly picking up where [others had left off](https://github.com/liscju/checkstyle/commit/d91467b132adda9c9fa216fb3b8c47074e4754e9) in issue #2974.

I completely agree that your suggested solution is the correct approach. I just don't have any time to dedicate to this at the moment. It's unfortunate, because I really would like this issue fixed.

**Comment by romani** [comment]
too much mess, we need to drop this PR and start from scratch.

lets close  this PR, I already closed issue with reason https://github.com/checkstyle/checkstyle/issues/2974#issuecomment-401198949, to drop all connection to previous fixes.

@jjlharrison , you are welcome with better PR for newly opened issue.
